include/wil/network.h

```diff
@@ -0,0 +1,1328 @@
+
+        [[nodiscard]] ADDRESS_FAMILY family() const WI_NOEXCEPT;
+        [[nodiscard]] USHORT port() const WI_NOEXCEPT;
+        [[nodiscard]] ULONG flow_info() const WI_NOEXCEPT;
+        [[nodiscard]] ULONG scope_id() const WI_NOEXCEPT;
+
+        [[nodiscard]] SOCKADDR* sockaddr() WI_NOEXCEPT;
+        [[nodiscard]] SOCKADDR_IN* sockaddr_in() WI_NOEXCEPT;
+        [[nodiscard]] SOCKADDR_IN6* sockaddr_in6() WI_NOEXCEPT;
+        [[nodiscard]] SOCKADDR_INET* sockaddr_inet() WI_NOEXCEPT;
+        [[nodiscard]] IN_ADDR* in_addr() WI_NOEXCEPT;
+        [[nodiscard]] IN6_ADDR* in6_addr() WI_NOEXCEPT;
+
+        [[nodiscard]] const SOCKADDR* sockaddr() const WI_NOEXCEPT;
+        [[nodiscard]] const SOCKADDR_IN* sockaddr_in() const WI_NOEXCEPT;
+        [[nodiscard]] const SOCKADDR_IN6* sockaddr_in6() const WI_NOEXCEPT;
+        [[nodiscard]] const SOCKADDR_INET* sockaddr_inet() const WI_NOEXCEPT;
+        [[nodiscard]] const IN_ADDR* in_addr() const WI_NOEXCEPT;
+        [[nodiscard]] const IN6_ADDR* in6_addr() const WI_NOEXCEPT;
+
+        [[nodiscard]] SOCKADDR_STORAGE sockaddr_storage() const WI_NOEXCEPT;
+
+        [[nodiscard]] constexpr int length() const
+        {
+            return sizeof(m_sockaddr);
+        }
+
+    private:
+        SOCKADDR_INET m_sockaddr{};
+    };
+
+    // When using dual-mode sockets, one might need to connect to a target IPv4 address.
+    // Since dual-mode socket types are IPv6, one must map that IPv4 address to its 'mapped' IPv6 address
+    inline ::wil::network::socket_address map_dual_mode_4to6(const ::wil::network::socket_address& ipv4_address) WI_NOEXCEPT
+    {
+        constexpr IN6_ADDR ipv4MappedPrefix{{IN6ADDR_V4MAPPEDPREFIX_INIT}};
+
+        ::wil::network::socket_address return_ipv6_address{&ipv4MappedPrefix, ipv4_address.port()};
+
+        auto* const return_in6_addr{return_ipv6_address.in6_addr()};
+        const auto* const ipv4_inaddr{ipv4_address.in_addr()};
+        return_in6_addr->u.Byte[12] = ipv4_inaddr->S_un.S_un_b.s_b1;
+        return_in6_addr->u.Byte[13] = ipv4_inaddr->S_un.S_un_b.s_b2;
+        return_in6_addr->u.Byte[14] = ipv4_inaddr->S_un.S_un_b.s_b3;
+        return_in6_addr->u.Byte[15] = ipv4_inaddr->S_un.S_un_b.s_b4;
+
+        return return_ipv6_address;
+    }
+
+    //! non-member swap
+    inline void swap(::wil::network::socket_address& lhs, ::wil::network::socket_address& rhs) WI_NOEXCEPT
+    {
+        lhs.swap(rhs);
+    }
+
+    //! class addr_info encapsulates the ADDRINFO-related structures returned from the socket functions
+    //! getaddrinfo, GetAddrInfoW, GetAddrInfoWEx
+    //! iterator semantics are supported to safely access these addresses
+    // ! template T supports pointers to the 3 address structures: ADDRINFOA*, ADDRINFOW*, ADDRINFOEXW*
+    template <typename T>
+    class addr_info_iterator_t
+    {
+    public:
+        static constexpr addr_info_iterator_t end() WI_NOEXCEPT

```

Comment:
Why have `end` with no `begin`? If you're going for a `std::filesystem::directory_iterator`-like design, you'll need both. E.g. for
```c++
for (auto xxx : add_info_iterator_t(...)) { ... }
```